src/net/sourceforge/kolmafia/textui/langserver/Script.java

```diff
@@ -0,0 +1,193 @@
+
+  /**
+   * The {@link Handler} in charge of this script. <i>Only</i> exists if this script is <b>not</b>
+   * imported by any other script.
+   *
+   * <p>When trying to find the handler(s) in charge of a given file, {@link
+   * FilesMonitor#findHandlers(File)} should be used.
+   */
+  Handler handler;
+
+  Script(final AshLanguageServer parent, final File file) {
+    this.parent = parent;
+    this.file = file;
+  }
+
+  Handler makeHandler() {
+    this.handler = new Handler();
+
+    this.parent.executor.execute(
+        () -> {
+          this.handler.parseFile();
+        });
+
+    return this.handler;
+  }
+
+  /**
+   * An object tasked with taking care of a {@link Script}, such as parsing it and sending the
+   * resulting diagnostics to the client once the parsing ends.
+   *
+   * <p>All files imported by this script should also be handled by this object
+   */
+  public class Handler {
+    Parser parser;
+    Scope scope;
+
+    private Thread parserThread;
+
+    private final Object parserSwapLock = new Object();
+    private final Object parserThreadWaitingLock = new Object();
+
+    // Public class, private constructor
+    private Handler() {}
+
+    private void parseFile() {
+      final String previousThreadName = Thread.currentThread().getName();
+
+      synchronized (this.parserSwapLock) {
+        this.parserThread = Thread.currentThread();
+        this.parserThread.setName(Script.this.file.getName() + " - Parser");
+      }
+
+      this.parser =
+          new LSParser(Script.this.file, null, Collections.synchronizedMap(new HashMap<>()));
+
+      try {
+        // Parse the script
+        this.scope = this.parser.parse();
+
+        // If we managed to parse it without interruption, send the diagnostics
+        Script.this.parent.executor.execute(
+            () -> {
+              this.sendDiagnostics();
+            });
+      } catch (InterruptedException e) {
+      } finally {
+        synchronized (this.parserSwapLock) {
+          if (this.parserThread == Thread.currentThread()) {
+            this.parserThread = null;
+
+            synchronized (this.parserThreadWaitingLock) {
+              this.parserThreadWaitingLock.notifyAll();
+            }
+          }
+
+          Thread.currentThread().setName(previousThreadName);
+        }
+      }
+    }
+
+    void sendDiagnostics() {
+      if (Script.this.parent.client == null) {
+        return;
+      }
+
+      this.waitForParsing();
+
+      if (Script.this.handler != this || this.parser == null || Thread.interrupted()) {
+        // We've been kicked out
+        return;
+      }
+
+      final String previousThreadName = Thread.currentThread().getName();
+      // Will technically send the diagnostics of every file it imports, but
+      // don't change the name accordingly; it would change too fast.
+      Thread.currentThread().setName(Script.this.file.getName() + " - Diagnostics");
+
+      synchronized (this.parser.getImports()) {
```

Comment:
synchronizing on a mutable map? Could something go wrong here?

Reply:
Given that we're iterating on that map's `entrySet` on the next line, yes, we want to prevent someone from accessing and modifying the map while we're doing this (Not sure "why" they would alter `parser.imports`, but it's just good practice)

Reply:
Sorry for the lack of precision here: I meant "is it possible that `this.parser.getImports()` could return two different things for two different threads?".

I've checked, and it seems that `parser.imports` is only assigned in the constructor. So I think this is fine.

Reply:
Just to clarify: I thought what you were originally saying was "I don't get why we need to synchronize here, what's the use?"

If I understand correctly, what you meant was: "I see we synchronize on *this*. Are we certain that *this* will always be what we think?"
Am I correct?

Reply:
Pretty much. More precisely, it's 

"I see we're synchronising on the output of a function. Are we sure the output of the function won't be different objects in different threads?"